rllib/env/multi_agent_episode.py

```diff
@@ -0,0 +1,718 @@
+import numpy as np
+import uuid
+
+from typing import Any, Dict, List, Optional, Union
+
+from ray.rllib.env.single_agent_episode import SingleAgentEpisode
+from ray.rllib.policy.sample_batch import MultiAgentBatch
+from ray.rllib.utils.typing import MultiAgentDict
+
+
+# TODO (simon): Include cases in which the number of agents in an
+# episode are shrinking or growing during the episode itself.
+class MultiAgentEpisode:
+    """Stores multi-agent episode data.
+
+    The central attribute of the class is the timestep mapping
+    `global_t_to_local_t` that maps the global (environment)
+    timestep to the local (agent) timesteps.
+
+    The `MultiAgentEpisode` is based on the `SingleAgentEpisode`s
+    for each agent, stored in `MultiAgentEpisode.agent_episodes`.
+    """
+
+    def __init__(
+        self,
+        id_: Optional[str] = None,
+        agent_episode_ids: Optional[Dict[str, str]] = None,
+        *,
+        observations: Optional[List[MultiAgentDict]] = None,
```

Comment:
I think what you said on slack is correct:


Simon:
```
But was with the case The environment is not terminated and some agents are terminated earlier on, e.g. timestep 10 agent 1 temrinated (because it fell down a cliff) the other agents are still alive at timestep 100. Do we need here probably in the _init_  of the MultiAgentEpisode`  is_terminated and is_terminated as a list of `MultiAgentDict`s like [{...}, ... , {"agent_1": True, "agent_2": False, ...}, ....] such that we know which agent is still alive and where agent 1 died?

Sven:
```
Great point! Maybe we should NOT allow users to construct a MA Episode via observations and actions and ... But rather by providing a bunch of agentIDs mapping to SingleAgentEpisodes?
```

Reply:
Just to keep this in mind: Thus far, initializing a new episode (single- or multi-agent) with already existing data is an edge case, that we have - to the best of my knowledge - never used in RLlib. So it's not a super urgent decision we need to make, just to keep this in the back of our heads.

Reply:
See my comment to this on slack. There might be a hurdle with MA episodes.